### DISCUSSION TITLE
How to check programmatically if Sumatra is installed?

### DISCUSSION THREAD

**Discussion by PHPirates** [sumatrapdfreader/sumatrapdf#2855] [Q&A] [answered]
We are developing a LaTeX IDE and are using DDE commands to communicate with Sumatra.
The past few years, we've relied on some registry key to check if Sumatra is installed (to show it in the list of available pdf viewers), but recently this stopped working.

What would be the best way for us to check if Sumatra is installed?
We are working in Kotlin (Java), it should not require user interaction.

**Comment by GitHubRulesOK** [comment, marked as answer]
the registry keys should be there it is only the single user installs that do not need to be registered so for example the last entry in current log is
HKEY_CURRENT_USER, Software\Classes\MIME\Database\Content Type\application/pdf, Extension, .pdf
but a more natural one earlier is
HKEY_CURRENT_USER, Software\Classes\SumatraPDF.pdf\Application, ApplicationName, SumatraPDF
or perhaps better
HKEY_LOCAL_MACHINE, Software\Classes\.pdf\OpenWithProgids, SumatraPDF

They can vary by version and registered file types so machine or user pdf is the most relevant for LaTeX

remember that DDE service is not available until exe is first invoked (i.e. not detectable before exe run time) so again a good location to look for the exe is like many apps the uninstall location 

HKEY_CURRENT_USER, Software\Microsoft\Windows\CurrentVersion\Uninstall\SumatraPDF, DisplayIcon, c:\program files\SumatraPDF\SumatraPDF.exe

However note that many users dont need to install that is only needed for windows search and pre-viewer. so in such cases the user may (or may not) assign pdfviewer by right click.

**Reply by PHPirates** [reply]
Thanks! But from what I'm seeing in my own registry, these keys are only present if you select `Install for all users` during installation, right? How do we go about detecting other installations?

**Reply by GitHubRulesOK** [reply]
If you installed then your likely to have Computer\HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Explorer\FileExts\.pdf\UserChoice
However Microsoft will often pull the rug from under other apps and set that for them to Edge

There is no sure fired way to know if I am using a particular SumatraPDF until either I point to whichever version I need or Microsoft has stored one of my many installs but then I tend to not be a "normal" installer

Simplest is find the exe in %userprofile%\AppData\Local\SumatraPDF and you can verify that's installed

```
[HKEY_CURRENT_USER\Software\Classes\SumatraPDF.pdf\shell\open]
"Icon"="C:\\Users\\K\\AppData\\Local\\SumatraPDF\\SumatraPDF.exe"

[HKEY_CURRENT_USER\Software\Classes\SumatraPDF.pdf\shell\open\command]
@="\"C:\\Users\\K\\AppData\\Local\\SumatraPDF\\SumatraPDF.exe\" \"%1\""
```

**Reply by PHPirates** [reply]
Ah okay, if checking for the default location installation is the best we can do in that case, we'll implement that. Many thanks for your help!
